server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregationBuilder.java

```diff
@@ -135,7 +136,22 @@ public boolean supportsSampling() {
     }
 
     @Override
-    public boolean supportsParallelCollection() {
+    public boolean supportsParallelCollection(ToLongFunction<String> fieldCardinalityResolver) {
+        // TODO unconditionally enable parallelism when ordered by key instead of count as precision is not affected
+        /*
+        if (InternalOrder.isKeyOrder(order)) {
```

Comment:
Given that we already take into account this in `adjustBucketCountThresholds()`, maybe we should just try to enable this in this PR. Assuming that it doesn't require more changes than this.

Reply:
Right. I am wondering if we should still disable concurrency for high cardinality fields, despite ordered by term. We would do the same for the composite agg as we've seen performance overhead in that case, caused by concurrency.

Reply:
Good point. Let's always take cardinality into account. Maybe remove this code comment and instead add a comment that in case of key order there is no accuracy issue, but we should always take cardinality into account to avoid a performance issue?

Reply:
will do so.

Reply:
Now that I updated the PR to always take cardinality into account, I wonder if we should look at shard_size when order is by term. Should we rather hardcode some threshold? We are going to have to do the same for composite as well.

Reply:
@martijnvg do you have thoughts about this?

Reply:
I think parallel collection should be disabled in case of high cardinality? (to avoid performance degradation)
I think in that case we need to always check the cardinality of a field. Then I think we just need to remove `InternalOrder.isKeyOrder(order) == false` in the `adjustBucketCountThresholds(...)` method?

Reply:
But is shard size applied when terms agg is ordered by term? I am thinking that it may make sense to have a different threshold, independent of shard_size, possibly hard-coded, when order is by term.

Reply:
> But is shard size applied when terms agg is ordered by term? 

I think I was confused. Shard size if increased only if not sorting by key.

>  I am thinking that it may make sense to have a different threshold, independent of shard_size, possibly hard-coded, when order is by term.

Are you referring to a threshold that would allow parallel collection when ordering by key but does not increase the shard size? An upper limit to avoid potential performance issues? I think this makes sense.

Reply:
> Are you referring to a threshold that would allow parallel collection when ordering by key but does not increase the shard size? An upper limit to avoid potential performance issues? I think this makes sense.

Yes that is what I meant, it looks like using shard_size as a threshold when ordering by key makes little sense. What would be a good place to start? 50? 100?

Reply:
I used 50 for now.

Reply:
👍